synapse/storage/databases/main/roommember.py

```diff
@@ -1019,14 +1040,26 @@ async def _check_host_room_membership(
 
     @cached(iterable=True, max_entries=10000)
     async def get_current_hosts_in_room(self, room_id: str) -> Set[str]:
-        """Get current hosts in room based on current state."""
+        """
+        Get current hosts in room based on current state.
+
+        The heuristic of sorting by servers who have been in the room the
+        longest is good because they're most likely to have anything we ask
+        about.
+
+        Returns:
+            Returns a list of servers sorted by longest in the room first. (aka.
+            sorted by join with the lowest depth first).
+        """
 
         # First we check if we already have `get_users_in_room` in the cache, as
         # we can just calculate result from that
         users = self.get_users_in_room.cache.get_immediate(
             (room_id,), None, update_metrics=False
         )
         if users is not None:
+            # Because `users` is sorted from lowest -> highest depth, the set of
+            # domains will also be sorted that way.
```

Comment:
This is not true; sets aren't ordered collections and their elements don't have any prescribed order.

(Python gives a misleading impression because CPython's *dicts* are ordered based on insertion order, but that's because of a CPython implementation detail that it uses a set of entries alongside a linked list)

As an example:
```
>>> list(set([2, 1]))
[1, 2]
```

You'll have to deduplicate some other way if the ordering is important

Reply:
Perhaps https://grantjenks.com/docs/sortedcollections/orderedset.html ? We already use SortedDict and SortedList elsewhere in Synapse

Reply:
ohhh, good shout! It looks like I also got lucky in my [other example](https://github.com/matrix-org/synapse/pull/13575/files#r952081093). Running it a few more times and I can see the ordering change around 😵

I guess we will have to return a `List[str]` here for the sort guarantee for this function in general

Reply:
Updated to use deduplicated `List[str]` 👍

Can switch to an ordered set if we decide to use that ⏩

Reply:
as an aside, confusing as that grantjenks has both 'sorted collections' and 'sorted containers' and they have different data structures...